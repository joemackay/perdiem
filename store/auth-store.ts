import * as SecureStore from 'expo-secure-store';
import { create } from 'zustand';
import { persist } from 'zustand/middleware';

const TOKEN = 'token';
const isTestEnv = process.env.JEST_WORKER_ID !== undefined;

interface AuthState {

  // unique token retrieved from the server
  token: string | null;

  // check if user has gone idle - a nice to have
  status: 'idle' | 'signOut' | 'signIn';
  getToken: () => Promise<string | null>;
  saveToken: (token: string) => void;
  clearToken: () => void;
}

export const _useAuth = create<AuthState>()(
  persist(
    (set, get) => ({
      status: 'idle',
      token: null,

      // fetch token for the provider
      getToken: async () => {
         const token = await SecureStore.getItemAsync(TOKEN)
         return token;
      },

      // save token in Secure store
      saveToken: async (token) => {
        // setToken(token);
        await SecureStore.setItemAsync(TOKEN, token)
        if (!isTestEnv) {
          set({ status: 'signIn', token });
        }
      },
      
      // Clear token from Secure store - User has logged out
      clearToken: async () => {
        // removeToken();
        console.log('3. AuthState clearToken')
        await SecureStore.deleteItemAsync(TOKEN)
        set({ status: 'signOut', token: null });
      }
    }),
    { 
      name: 'auth-storage' 
    }
));