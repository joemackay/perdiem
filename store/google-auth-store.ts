// // src/auth/googleAuth.ts
// import { useAuthRequest } from 'expo-auth-session';
// import { GoogleAuthProvider, signInWithCredential } from 'firebase/auth';
// import { useEffect } from 'react';
// import { auth } from './firebase';

// // Your Google Client ID
// const CLIENT_ID = 'YOUR_GOOGLE_CLIENT_ID.apps.googleusercontent.com';

// export const useGoogleAuth = () => {
//   const [request, response, promptAsync] = useAuthRequest({
//     clientId: CLIENT_ID,
//     scopes: ['profile', 'email'],
//   });

//   useEffect(() => {
//     if (response?.type === 'success' && response.params?.id_token) {
//       const { id_token } = response.params;

//       // Create a Google credential with the ID token
//       const credential = GoogleAuthProvider.credential(id_token);

//       // Sign in with Firebase using the Google credential
//       signInWithCredential(auth, credential)
//         .then((userCredential) => {
//           console.log('User signed in:', userCredential.user);
//         })
//         .catch((error) => {
//           console.error('Error signing in with Google:', error);
//         });
//     }
//   }, [response]);

//   return {
//     request,
//     promptAsync,
//   };
// };
