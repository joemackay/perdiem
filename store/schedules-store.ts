// stores/authStore.ts
import { StoreScheduleType } from '@/types/storescheduletime';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';

// Store the state of time schedules of the shop
type ScheduleState = {
  timezone: string | null;
  selectedDayOfTheMonth: number | null;
  selectedTime: string | null;
  storeSchedules: StoreScheduleType[];
  storeScheduleOverides: StoreScheduleType[];
  setSelectedTimezone: (tz: string) => void;
  setSelectedDayOfTheMonth: (date: number) => void,
  setSelectedTime: (time: string) => void,
  setSchedules: (schedules: StoreScheduleType[]) => void,
  setScheduleOverides: (schedules: StoreScheduleType[]) => void
};

export const useSchedulesStore = create<ScheduleState>()(
  persist(
    (set) => ({
      //the default timezone
      timezone: null,

      // number format day of the week
      selectedDayOfTheMonth: null,

      // store selected time for use in other pages
      selectedTime: null,

      // store the list of schedules 
      storeSchedules: [],

      // store the list of schedule overides
      storeScheduleOverides: [],
      setSelectedTimezone: (tz: string) => {
        set({ timezone: tz })
      },
      setSelectedDayOfTheMonth: (date: number) => {
        set({ selectedDayOfTheMonth: date })
      },
      setSelectedTime: (time: string) => {
        set({ selectedTime: time })
      },
      setSchedules: (schedules: StoreScheduleType[]) => {
        set({storeSchedules : schedules})
      },
      setScheduleOverides: (schedules: StoreScheduleType[]) => {
        set({storeScheduleOverides : schedules})
      }
    }),
    { name: 'schedules-storage',
      storage: createJSONStorage(() => AsyncStorage), // Explicitly use AsyncStorage 
      }
  )
);
