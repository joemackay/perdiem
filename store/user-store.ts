// stores/authStore.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { create } from 'zustand';
import { createJSONStorage, persist } from 'zustand/middleware';
import { User } from '../types/user';

// Store the state of the user auth state
type UserState = {
  user: User | null;
  setUser: (user: User) => void;
  getUser: () => User | null;
  logout: () => void;
};

export const useUserStore = create<UserState>()(
  persist(
    (set, get) => ({

      // User record
      user: null,

      // Store user record
      setUser: (user) => set({ user }),

      // Get  user record
      getUser: () => {
        const state = get();
        return state.user;
      },

      // Clear user record
      logout: () => {
        console.log('2. useUserStore logout')
        set({ user: null })
      },
    }),
    { 
      name: 'auth-storage',
      storage: createJSONStorage(() => AsyncStorage), // Explicitly use AsyncStorage
    }
  )
);
