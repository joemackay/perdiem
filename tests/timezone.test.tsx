import { useSchedulesStore } from '@/store/schedules-store';
import TimezoneToggle from '@/ui/TimeZoneToggle';
import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

// Mock the Zustand store
jest.mock('@/store/schedules-store');

describe('TimezoneToggle', () => {
  const mockSetSelectedTimezone = jest.fn();
  const mockLocale = 'Europe/London';

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Mock the schedules store
    (useSchedulesStore as unknown as jest.Mock).mockImplementation((selector) => {
      if (selector === jest.fn()) {
        return {
          timezone: 'America/New_York',
          setSelectedTimezone: mockSetSelectedTimezone,
        };
      }
      return {};
    });
  });

  it('renders correctly with initial props', () => {
    const { getByTestId, getByText } = render(
      <TimezoneToggle locale={mockLocale} />
    );

    // Check container and elements are rendered
    expect(getByTestId('test-tz-container')).toBeTruthy();
    expect(getByTestId('test-tz-locale')).toBeTruthy();
    expect(getByTestId('test-tz-switch')).toBeTruthy();
    
    // Check text content
    expect(getByText(mockLocale)).toBeTruthy();
    expect(getByText('NYC Timezone')).toBeTruthy();
  });

  it('displays switch in correct position based on timezone', () => {
    // Test when timezone is NYC
    (useSchedulesStore as unknown as jest.Mock).mockImplementation((selector) => ({
      timezone: 'America/New_York',
      setSelectedTimezone: mockSetSelectedTimezone,
    }));
    
    const { getByTestId, rerender } = render(
      <TimezoneToggle locale={mockLocale} />
    );
    
    const switchElement = getByTestId('test-tz-switch');
    expect(switchElement.props.value).toBe(true);

    // Test when timezone is not NYC
    (useSchedulesStore as unknown as jest.Mock).mockImplementation((selector) => ({
      timezone: 'Europe/London',
      setSelectedTimezone: mockSetSelectedTimezone,
    }));
    
    rerender(<TimezoneToggle locale={mockLocale} />);
    expect(switchElement.props.value).toBe(false);
  });

  it('calls setSelectedTimezone with correct values when toggled', () => {
    const { getByTestId } = render(
      <TimezoneToggle locale={mockLocale} />
    );

    const switchElement = getByTestId('test-tz-switch');
    
    // First toggle (from NYC to locale)
    fireEvent(switchElement, 'valueChange', false);
    expect(mockSetSelectedTimezone).toHaveBeenCalledWith(mockLocale);
    
    // Second toggle (back to NYC)
    fireEvent(switchElement, 'valueChange', true);
    expect(mockSetSelectedTimezone).toHaveBeenCalledWith('America/New_York');
  });

  it('updates internal state when toggled', () => {
    const { getByTestId } = render(
      <TimezoneToggle locale={mockLocale} />
    );

    const switchElement = getByTestId('test-tz-switch');
    
    // Initial state (true for NYC timezone)
    expect(switchElement.props.value).toBe(true);
    
    // Toggle off
    fireEvent(switchElement, 'valueChange', false);
    expect(switchElement.props.value).toBe(false);
    
    // Toggle on
    fireEvent(switchElement, 'valueChange', true);
    expect(switchElement.props.value).toBe(true);
  });

  it('matches snapshot', () => {
    const { toJSON } = render(
      <TimezoneToggle locale={mockLocale} />
    );
    
    expect(toJSON()).toMatchSnapshot();
  });
});