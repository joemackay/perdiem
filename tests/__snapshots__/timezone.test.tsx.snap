// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimezoneToggle matches snapshot 1`] = `
<View
  className="bg-white flex-row justify-left items-center"
  testID="test-tz-container"
>
  <Text
    className="mr-2 text-black"
    testID="test-tz-locale"
  >
    Europe/London
  </Text>
  <RCTSwitch
    accessibilityRole="switch"
    onChange={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "height": 31,
        "width": 51,
      }
    }
    testID="test-tz-switch"
    value={false}
  />
  <Text
    className="ml-2"
  >
    NYC Timezone
  </Text>
</View>
`;
